[Cache] Allow and use generators in AbstractAdapter

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

The most important enhancement is allowing doFetch to return a generator, which wasn't possible before.
The other changes add strictness.

nicolas-grekas committed:
[Cache] Allow and use generators in AbstractAdapter

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -80,9 +80,11 @@ function ($deferred, $namespace) {
      * @return bool True if the pool was successfully cleared, false otherwise.
      */
-    abstract protected function doClear();
+    abstract protected function doClear($namespace);
```
I propose adding $namespace to this abstract method so that adapters can flush only a subpart of their global pool if that applies

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -114,7 +116,7 @@ public function getItem($key)
         }
         if (isset($this->deferred[$key])) {
-            return $this->deferred[$key];
+            return clone $this->deferred[$key];
```
bug fix

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -165,11 +159,20 @@ public function getItems(array $keys = array())
+        if (isset($this->deferred[$key])) {
+            static $prefix = "\0Symfony\Component\Cache\CacheItem\0";
+            $item = (array) $this->deferred[$key];
+            $ok = $this->doSave(array($item[$prefix.'key'] => $item[$prefix.'value']), $item[$prefix.'lifetime']);
```
optimization: don't commit all deferred items on a hasItem

nicolas-grekas commented on src/Symfony/Component/Cache/CacheItem.php:
```diff
@@ -19,6 +19,11 @@
+    /**
+     * @internal
+     */
+    const CAST_PREFIX = "\0Symfony\Component\Cache\CacheItem\0";
```
ping @dunglas

dunglas replied:
:+1:

stof commented:
@nicolas-grekas is this subject to race conditions between reading items and updating the cache during the iteration ?

nicolas-grekas commented:
@stof I'd say it's up to the concrete adapter to deal with this. If one adapter chooses to return an iterator/generator, it's up to it to it to deal with this kind of consequences. Do you agree?

stof commented:
@nicolas-grekas even if they don't return a generator, the cache system still use one, and performs ``doFetch`` in the middle of the generator iteration. So it may also be impacted.

nicolas-grekas commented:
I they do return an array (like Apcu/Doctrine adapters) then the result set is "frozen" in the generator. The behavior is well defined, so not subject to race conditions. PSR-6 doesn't specify if returned traversable are "live" or not, but we could argue that the "array" requirement means they should be frozen. We maybe should add this in our test suite, don't you think?

stof commented:
@nicolas-grekas adding this to the testsuite is a good idea (and is what I suggested in https://github.com/php-cache/integration-tests/pull/34#issuecomment-172880337)

Are you sure that adapters cannot be affected ? doFetch is not at the beginning of the iteration

nicolas-grekas commented:
@stof I moved the doFetch first and also added an isset check around deferred.
About deferred, this should be always empty since we commit them before fetching.
$deferred can still be non empty if saving some deferred failed.
Could you review the failure handling logic around deferred in the while? It's a complex topic and you may have some ideas for the better.

nicolas-grekas commented:
@stof btw adding the test we're talking about should maybe not be included in the integration tests but only in our own test suite, since psr-6 has no requirements here (unless someone explains that there are?)

nicolas-grekas commented on src/Symfony/Component/Cache/composer.json:
```diff
@@ -23,7 +23,7 @@
     },
     "require-dev": {
-        "cache/integration-tests": "^0.6",
+        "cache/integration-tests": "dev-master",
```
We need this until the test suite fits our needs. I suggest we add a version number when the Component is stable enough/feature full, in a few weeks.

nicolas-grekas commented on composer.json:
```diff
@@ -76,7 +76,7 @@
     },
     "require-dev": {
-        "cache/integration-tests": "^0.6",
+        "cache/integration-tests": "dev-master",
```
We need this until the test suite fits our needs. I suggest we add a version number when the Component is stable enough/feature full, in a few weeks.

derrabus replied:
This could be changed to `"^0.7"` now, I guess?

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -230,10 +223,7 @@ public function saveDeferred(CacheItemInterface $item)
         }
         try {
             $item = clone $item;
-        } catch (\Error $e) {
```
I propose to not deal with \Error here

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -259,13 +248,22 @@ public function commit()
+        $this->deferred = array();
+        $ok = true;
+
+        // When bulk-save failed, retry each item individually
```
The previous logic has a "memory leak" in that whenever a deferred couldn't saved, it would never be freed.
The logic I propose here is: try bulk-save, then try individual save, then trash.

nicolas-grekas commented on src/Symfony/Component/Cache/Adapter/AbstractAdapter.php:
```diff
@@ -259,13 +248,28 @@ public function commit()
+                    $ok = false;
+
+                    foreach ($v as $key => $value) {
+                        @trigger_error(sprintf('Failed to cache key "%s" of type "%s"', is_object($value) ? get_class($value) : gettype($value)));
```
This could help someone someday figuring out what's wrong...

nicolas-grekas commented:
@stof PR is ready. Please note that I'm now AFK until the end of January, so feel free to take over this PR as I won't be able to resolve any comment until then.

nicolas-grekas commented:
No comment? Ping @symfony/deciders

fabpot commented:
Thank you @nicolas-grekas.
